internal/cmd/util/validation.go

```diff
@@ -0,0 +1,44 @@
+package util
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var posArgRegex = regexp.MustCompile(`\[.*?]|\(.*?\)|--.*?<.*?>|<(.*?)>`) // matches all positional args in group 1
+
+func ValidateExact(cmd *cobra.Command, args []string) error {
+	return validate(cmd, args, true)
+}
+
+func Validate(cmd *cobra.Command, args []string) error {
+	return validate(cmd, args, false)
+}
+
+func validate(cmd *cobra.Command, args []string, exact bool) error {
+	matches := posArgRegex.FindAllStringSubmatch(cmd.Use, -1)
+	var expected []string
+	for _, match := range matches {
+		if match[1] != "" {
+			expected = append(expected, match[1])
+		}
+	}
+	if len(args) > len(expected) && exact {
+		return fmt.Errorf("expected exactly %d positional arguments, but got %d", len(expected), len(args))
+	}
+	for i := 0; i < len(expected); i++ {
+		if i >= len(args) {
+			idx := strings.Index(cmd.Use, "<"+expected[i]+">")
+			if idx != -1 {
+				_, _ = fmt.Fprintln(os.Stderr, cmd.Use)
```

Comment:
I love the helpful error messages here! I have one suggestion, but if its too hard to implement or you would prefer to do it in a follow up PR that is fine by me :)

Right now this prints something like:

```shell
➜  hcloud server delete
delete <server>
        ^^^^^^
hcloud: expected argument server at position 1
```

I think the output could be a bit clearer if it would include all parent commands, not only the current command we are validating:

```shell
➜  hcloud server delete
hcloud server delete <server>
                      ^^^^^^
hcloud: expected argument server at position 1
```

Reply:
Yea, I was suprised about the `hcloud: %w` error output.